### ISSUE TITLE
[Bug] `no-unreachable` is not thrown

### ISSUE THREAD

**Issue by wenfangdu** [typescript-eslint/typescript-eslint#3583]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

[Repo Link](https://github.com/wenfangdu/vscode-jump/blob/no-unreachable-repro/src/extension.ts)

```TS
const fn = () => {
  return
  // 👇 should've thrown `no-unreachable`
  console.log()
}
```

**Expected Result**

`no-unreachable` is thrown.

**Actual Result**

No error is thrown. But if I comment out `'plugin:@typescript-eslint/recommended'`, `no-unreachable` will be thrown.

**Versions**

| package                            | version   |
| ---------------------------------- | --------- |
| `@typescript-eslint/eslint-plugin` | `v4.28.0` |
| `@typescript-eslint/parser`        | `v4.28.0` |
| `TypeScript`                       | `v4.3.4`  |
| `ESLint`                           | `v7.29.0` |
| `node`                             | `v16.3.0` |

**Comment by bradzacher** [comment]
This is intentional.
Typescript has options (that you should turn on) that do unreachable code detection better than ESLint ever could.

**Comment by wenfangdu** [comment]
@bradleyayers I've added `"allowUnreachableCode": false` to `tsconfig.json`, right now, it's been shown as warnings, how to make it shown as errors?

![image](https://user-images.githubusercontent.com/28700378/123638790-24211d80-d852-11eb-973d-0028762f510f.png)

> This is intentional.
Typescript has options (that you should turn on) that do unreachable code detection better than ESLint ever could.

Also, where is this behavior documented? Would like to know other rules having this same situation.

**Comment by wenfangdu** [comment]
@bradleyayers I've added `"allowUnreachableCode": false` to `tsconfig.json`, right now, it's been shown as warnings, how to make it shown as errors?

![image](https://user-images.githubusercontent.com/28700378/123638790-24211d80-d852-11eb-973d-0028762f510f.png)

_UPDATE_ Set the `reportStyleChecksAsWarnings` in VS Code settings to `false` will do.

> This is intentional.
Typescript has options (that you should turn on) that do unreachable code detection better than ESLint ever could.

Is this behavior documented? Would like to know other rules having this same situation.
